File: robocode.tests.robots/src/main/java/tested/robots/RobotDeathEvents.java
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
package tested.robots;


import robocode.DeathEvent;
import robocode.RobotDeathEvent;
import robocode.AdvancedRobot;


/**
 * @author Flemming N. Larsen (original)
 */
public class RobotDeathEvents extends AdvancedRobot {
	private boolean dead;
	private long enemyCount;

	public void run() {
		enemyCount = getOthers();
		while (!dead) {
			if (enemyCount != getOthers()) {
				throw new RuntimeException("enemyCount != getOthers()");
			}
			execute();
		}
	}

	public void onRobotDeath(RobotDeathEvent e) {
		enemyCount--;
	}

	public void onDeath(DeathEvent e) {
		dead = true;
	}
}

File: plugins/dotnet/robocode.dotnet.robotscs/src/tested/robotscs/RobotDeathEvents.cs
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
using Robocode;
using Robocode.Exception;

namespace tested.robotscs
{
    public class RobotDeathEvents : AdvancedRobot
    {
        private bool dead;
        private long enemyCount;

        public override void Run()
        {
            enemyCount = Others;
            while (!dead)
            {
                if (enemyCount != Others)
                {
                    throw new RobotException("enemyCount != GetOthers()");
                }
                Execute();
            }
        }

        public override void OnRobotDeath(RobotDeathEvent e)
        {
            enemyCount--;
        }

        public override void OnDeath(DeathEvent e)
        {
            dead = true;
        }
    }
}